batcher.go

```diff
@@ -0,0 +1,123 @@
+package nakadi
+
+import (
+	"reflect"
+	"time"
+)
+
+// publishAPI defines interface that is used for publishing. Used because of unit tests
+type publishAPI interface {
+	Publish(events interface{}) error
+}
+
+// BatchPublisher is a proxy that allows publishing events in a batched manner. In case if events are published in
+// parallel (not in a form of slices, but in a form of events), then batcher will collect them into batches, respecting
+// batch collection timeout and max batch size and instead of creating many separate requests to nakadi it will create
+// only several of them with aggregated data.
+type BatchPublisher struct {
+	publishAPI       publishAPI
+	options          BatchPublisherOptions
+	eventsChannel    chan *eventToPublish
+	dispatchFinished chan int
+}
+
+// BatchPublisherOptions specifies parameters that should be used to collect events to batches
+type BatchPublisherOptions struct {
+	// Maximum amount of time that event will spend in intermediate queue before being published.
+	BatchCollectionTimeout time.Duration
+	// Maximum batch size - it is guaranteed that not more than MaxBatchSize events will be sent within one batch
+	MaxBatchSize int
+}
+
+// NewPublishingBatcher creates a proxy for batching based on api and batching parameters
+func NewPublishingBatcher(api *PublishAPI, options BatchPublisherOptions) *BatchPublisher {
```

Comment:
1. Please also rename the constructor to `NewBatchPublisher.
2. If you really don't think that there should not be defaults, you should at least check that the batch size is not Zero. Because in its current implementation a zero value for options (`BatchPublisherOptions{}`) acts like `MaxBatchSize` defaults to `1`. Solution: either check for `MaxBatchSize` > `0` or implement `withDefaults()` as suggested before (and document the defaults in the options struct).
3. Other constructors use pointers to options. The reason for this is that options can be nil if they have defaults. Although this is currently not the case for  `BatchPublisherOptions`, I'd still prefer if the type is a pointer in order to avoid API changes in the future when reasonable defaults are in place.